### PR TITLE
Index balances by address and tokenID

### LINKED ISSUE #1383: Index balances by address and tokenID
### Description

Enable wallet and third party explorers to view all the top accounts (per token) for given blockchain.

### Motivation

- Enable users/auditors to explore the top accounts 
- Enable users to view the balance per token (multi token blockchains)

### Additional Information
- Index balances from the genesis block assets by address and tokenID for a given blockchain at init
- Update balances based on `new_block` and `delete_block` WS events from the SDK

### PR DESCRIPTION
**Pull request by priojeetpriyom**
### What was the problem?
This PR resolves #1383

### How was it solved?
- [x] Index balance by address and tokenID on block events
- [x] Refactor `getTokenBalance` to return all token balances for an address when limit or offset is not passed
- [x] Move all queue configurations under config file 
- [x] Add test coverage

### How was it tested?
- [x] Local

### CODE REVIEW

File: services/blockchain-indexer/shared/dataService/business/token.js [hunk 1]
```diff
@@ -56,7 +56,9 @@ const getTokenBalances = async (params) => {
 		if (response.balances) tokensInfo.push(...response.balances);
 	}
 
-	tokens.data = tokensInfo.slice(params.offset, params.offset + params.limit);
+	tokens.data = params.offset && params.limit
```
Comment (vardan10) [inline services/blockchain-indexer/shared/dataService/business/token.js:59]: What if we just pass either limit or offset and not both? In that case we might be returning entire tokenInfo array.

Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/token.js:59]: True, wanted to keep the changes minimal, thus did not handle all the cases. When we invoke using gateway we will have the default values, modified this only to suite the case when we do not pass anything.

File: services/blockchain-indexer/shared/indexer/accountBalanceIndex.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	}));
+
+	// Update all token balances of the address
+	await accountBalancesTable.upsert(updatedTokenBalances);
+};
+
+const accountBalanceIndexQueue = Queue(config.endpoints.cache, 'accountBalanceIndexQueue', updateAccountBalances, 1);
+
+const scheduleAccountBalanceUpdateFromEvents = async (events) => {
+	const tokenModuleEvents = events.filter(event => event.module === MODULE.TOKEN);
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/accountBalanceIndex.js:55]: We can discuss offline to make the implementation more efficient.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-indexer/tests/functional/shared/indexer/acccountBalanceIndex.test.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
+jest.setTimeout(30000);
```
Comment (priojeetpriyom) [inline services/blockchain-indexer/tests/functional/shared/indexer/acccountBalanceIndex.test.js:1]: will be removed in next commit

File: services/blockchain-indexer/tests/functional/shared/indexer/acccountBalanceIndex.test.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+		await accountBalancesTable.delete({});
+		const balanceBeforeUpdate = await accountBalancesTable.find({ address: accountInfo.address, tokenID: accountInfo.tokenID }, ['balance']);
+		expect(balanceBeforeUpdate.length).toBe(0);
+
+		await updateAccountBalances(accountInfo.address);
+
+		// Check balance after update
+		const balanceAfterUpdate = await accountBalancesTable.find({ address: accountInfo.address, tokenID: accountInfo.tokenID }, ['balance']);
+		expect(balanceAfterUpdate.length).toBe(1);
+		expect(balanceAfterUpdate[0].balance).toBeGreaterThanOrEqual(0);
```
Comment (vardan10) [inline services/blockchain-indexer/tests/functional/shared/indexer/acccountBalanceIndex.test.js:86]: Should we validate the balance as well?

```suggestion

                // Get current account balance
                const accountBalanceBeforeUpdate = await accountBalancesTable.find({ address: accountInfo.address, tokenID: accountInfo.tokenID }, ['balance']);
		const initialBalance = accountBalanceBeforeUpdate[0].balance;

		// Delete all balances and check balance before update
		await accountBalancesTable.delete({});
		const balanceBeforeUpdate = await accountBalancesTable.find({ address: accountInfo.address, tokenID: accountInfo.tokenID }, ['balance']);
		expect(balanceBeforeUpdate.length).toBe(0);

		await updateAccountBalances(accountInfo.address);

		// Check balance after update
		const balanceAfterUpdate = await accountBalancesTable.find({ address: accountInfo.address, tokenID: accountInfo.tokenID }, ['balance']);
		expect(balanceAfterUpdate.length).toBe(1);
		expect(balanceAfterUpdate[0].balance).toBeGreaterThanOrEqual(initialBalance);
```

File: services/blockchain-indexer/tests/functional/shared/indexer/acccountBalanceIndex.test.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+		// Delete all balances and check balance before update
+		const accountInfo = usersSubStoreInfos[0];
+		await accountBalancesTable.delete({});
+		const balanceBeforeUpdate = await accountBalancesTable.find({ address: accountInfo.address, tokenID: accountInfo.tokenID }, ['balance']);
+		expect(balanceBeforeUpdate.length).toBe(0);
+
+		await scheduleAccountBalanceUpdateFromEvents(eventsIncludingTokenModule);
+
+		// Check balance after update
```
Comment (vardan10) [inline services/blockchain-indexer/tests/functional/shared/indexer/acccountBalanceIndex.test.js:110]: Should we update this comment?

File: services/blockchain-indexer/shared/indexer/accountBalanceIndex.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		address,
+		tokenID: balanceInfo.tokenID,
+		balance: balanceInfo.availableBalance,
+	}));
+
+	// Update all token balances of the address
+	await accountBalancesTable.upsert(updatedTokenBalances);
+};
+
+const accountBalanceIndexQueue = Queue(config.endpoints.cache, 'accountBalanceIndexQueue', updateAccountBalances, 1);
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/indexer/accountBalanceIndex.js]: Applied.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-indexer/shared/dataService/business/token.js [hunk 7]
```diff
@@ -68,6 +82,49 @@ const getTokenBalances = async (params) => {
...
+	const queryParams = {
+		tokenID: params.tokenID,
+	};
+
+	// Count total number of response before adding limit/offset
+	const totalAddressCount = await accountBalancesTable.count(queryParams);
+
+	if (params.limit) queryParams.limit = params.limit;
+	if (params.offset) queryParams.offset = params.offset;
+	queryParams.sort = params.sort ? params.sort : 'balance:desc';
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/token.js:100]: Please build the queryParams at once with all the necessary information. `count` method already ignores the pagination properties.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-indexer/shared/dataService/business/token.js (hunk 7 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/token.js]: Applied.

[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/token.js]: Applied.

File: services/blockchain-indexer/shared/dataService/business/token.js [hunk 8]
```diff
@@ -68,6 +82,49 @@ const getTokenBalances = async (params) => {
...
+	response.data[params.tokenID] = await BluebirdPromise.map(
+		tokenInfos,
+		async tokenInfo => {
+			const accountInfo = await getIndexedAccountInfo({ address: tokenInfo.address }, ['publicKey', 'name']);
+			return {
+				address: tokenInfo.address,
+				publicKey: accountInfo.publicKey,
+				name: accountInfo.name,
+				balance: tokenInfo.balance,
+				knowledge: null,
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/token.js]: Applied.

File: services/blockchain-indexer/shared/dataService/business/token.js [hunk 9]
```diff
@@ -152,9 +209,37 @@ const getTokenConstants = async () => {
...
+	const accountBalancesTable = await getAccountBalancesTable();
+
+	const tokenInfos = await accountBalancesTable.find({
+		distinct: 'tokenID',
+		sort: 'tokenID:asc',
+		offset: params.offset,
+		limit: params.limit,
+	},
+	['tokenID'],
+	);
```
[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/token.js]: Applied.

File: docs/api/version3.md [hunk 10]
```diff
@@ -1161,6 +1161,7 @@ Retrieves all the available token identifiers.
 | Parameter | Type | Validation | Default | Comment |
 | --------- | ---- | ---------- | ------- | ------- |
 | limit | Number | `[1,100]` | 10 |  |
+| sort | Enum | `["tokenID:desc", "tokenID:asc"]` | tokenID:asc |  |
 | offset | Number | `[0,Inf)` | 0 |  |
```
Comment (sameersubudhi) [inline docs/api/version3.md:1165]:
```suggestion
| sort | Enum | `["tokenID:desc", "tokenID:asc"]` | tokenID:asc |  |
| limit | Number | `[1,100]` | 10 |  |
| offset | Number | `[0,Inf)` | 0 |  |
```

File: services/blockchain-indexer/methods/dataService/modules/token.js [hunk 11]
```diff
@@ -72,6 +72,7 @@ module.exports = [
 		name: 'token.available-ids',
 		controller: getAvailableTokenIDs,
 		params: {
+			sort: { optional: true, type: 'string', default: 'tokenID:asc' },
```
Comment (sameersubudhi) [inline services/blockchain-indexer/methods/dataService/modules/token.js:75]:
```suggestion
			sort: { optional: true, type: 'string' },
```

File: tests/integration/api_v3/rpc/tokenAvailableIDs.test.js [hunk 12]
```diff
@@ -36,7 +36,25 @@ describe('get.token.available-ids', () => {
 		const response = await getTokensIDs({});
 		expect(response).toMap(jsonRpcEnvelopeSchema);
 		const { result } = response;
-		expect(result).toMap(goodResponseSchemaFortokenAvailableIDs);
+		expect(result).toMap(goodResponseSchemaForTokenAvailableIDs);
```
Comment (sameersubudhi) [inline tests/integration/api_v3/rpc/tokenAvailableIDs.test.js:39]: Could we also add conditions to check that the sorting is done properly?

File: tests/integration/api_v3/http/tokenAvailableIDs.test.js [hunk 13]
```diff
@@ -31,28 +31,42 @@ const endpoint = `${baseUrlV3}/token/available-ids`;
 describe('Token IDs API', () => {
 	it('should retrieve available token ids when called without any params', async () => {
 		const response = await api.get(endpoint);
-		expect(response).toMap(goodResponseSchemaFortokenAvailableIDs);
+		expect(response).toMap(goodResponseSchemaForTokenAvailableIDs);
```
Comment (sameersubudhi) [inline tests/integration/api_v3/http/tokenAvailableIDs.test.js:34]: Same as above.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-indexer/shared/dataService/business/token.js [hunk 14]
```diff
@@ -152,9 +207,38 @@ const getTokenConstants = async () => {
...
+	const accountBalancesTable = await getAccountBalancesTable();
+
+	const tokenInfos = await accountBalancesTable.find(
+		{
+			distinct: 'tokenID',
+			sort: params.sort,
+			offset: params.offset,
+			limit: params.limit,
+		},
+		['tokenID'],
```
Comment (vardan10) [inline services/blockchain-indexer/shared/dataService/business/token.js:224]:
```suggestion
                { ...params, distinct: 'tokenID' },
                ['tokenID'],
```

File: services/blockchain-indexer/shared/dataService/business/token.js [hunk 15]
```diff
@@ -68,6 +82,47 @@ const getTokenBalances = async (params) => {
...
+	};
+	const accountBalancesTable = await getAccountBalancesTable();
+	const queryParams = {
+		tokenID: params.tokenID,
+		sort: params.sort,
+		offset: params.offset,
+		limit: params.limit,
+	};
+
+	const tokenInfos = await accountBalancesTable.find(queryParams, ['address', 'balance']);
```
Comment (vardan10) [inline services/blockchain-indexer/shared/dataService/business/token.js:98]: Cant we directly pass params?
```suggestion
	const tokenInfos = await accountBalancesTable.find({...params}, ['address', 'balance']);
```

Comment (vardan10) [inline services/blockchain-indexer/shared/dataService/business/token.js:98]: Cant we directly pass params?
```suggestion
	const tokenInfos = await accountBalancesTable.find({...params}, ['address', 'balance']);
```

**Review by vardan10 — APPROVED** [review]

**Review by sameersubudhi — APPROVED** [review]

**Pull request merged**